### PR TITLE
Switch reverse PInvoke to the NativeCallable plan

### LINKED ISSUE #33581: Remove UMThunkStub for all non-x86 scenarios
With the completion of support for NativeCallableAttribute being public - https://github.com/dotnet/runtime/pull/33005, there is some additional cleanup that can now occur. The `UMThunkStub()` calls written in assembly can now be replaced by the newly added JIT helpers for Reverse P/Invoke Enter/Exit.

https://github.com/dotnet/runtime/blob/a1af0f2ad6cf51107a636dbe280cf939abd46b34/src/coreclr/src/vm/jithelpers.cpp#L5064-L5095

The x86 case must still use the `UMThunkStub()`, but all other can now be removed.

/cc @jkoritzinsky @jkotas @janvorli

### PR DESCRIPTION
**Pull request by jkotas**
Fixes #33581 (except x86)

### CODE REVIEW

File: src/coreclr/src/vm/jithelpers.cpp [hunk 1]
```diff
@@ -5397,13 +5403,17 @@ EXTERN_C void JIT_ReversePInvokeEnter(ReversePInvokeFrame* frame)
...
+
         // Manually inline the fast path in Thread::DisablePreemptiveGC().
         thread->m_fPreemptiveGCDisabled.StoreWithoutBarrier(1);
         if (g_TrapReturningThreads.LoadWithoutBarrier() == 0)
         {
-            frame->currentThread = thread;
             return;
         }
+
+        JIT_ReversePInvokeEnterRare2(frame);
```
Comment (jkotas) [inline src/coreclr/src/vm/jithelpers.cpp]: This is fixing a bug that was uncovered by this change. We are switched to cooperative mode here that lead to `ReversePInvokeBadTransition` being triggered.

File: src/coreclr/src/vm/dllimportcallback.h [hunk 2]
```diff
@@ -198,31 +184,33 @@ class UMThunkMarshInfo
...
 
-#if defined(TARGET_X86) && defined(FEATURE_STUBS_AS_IL)
+#ifdef FEATURE_STUBS_AS_IL
     struct ArgumentRegisters
     {
         UINT32 Ecx;
         UINT32 Edx;
     };
 
     VOID SetupArguments(char *pSrc, ArgumentRegisters *pArgRegs, char *pDst);
```
Comment (jkotas) [inline src/coreclr/src/vm/dllimportcallback.h]: This argument shuffling needs to be fixed to make the nativecallable path usable for non-Windows x86.

File: src/coreclr/src/vm/dllimportcallback.cpp [hunk 3]
```diff
@@ -1194,33 +1209,20 @@ VOID UMThunkMarshInfo::RunTimeInit()
 
     if (pFinalILStub == NULL)
     {
-        if (pMD != NULL && !pMD->IsEEImpl() &&
-            !NDirect::MarshalingRequired(pMD, GetSignature().GetRawSig(), GetModule()))
-        {
-            // Call the method directly in no-delegate case if possible. This is important to avoid JITing
-            // for stubs created via code:ICLRRuntimeHost2::CreateDelegate during coreclr startup.
-            pFinalILStub = pMD->GetMultiCallableAddrOfCode();
-        }
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/src/vm/dllimportcallback.cpp]: Can you unpack this issue for me? If the `MethodDesc` exists, but it wasn't implemented by the runtime and it is a blittable signature, just return the stub? Is this no longer interesting or am I missing why this is no longer a scenario.

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs [hunk 4]
```diff
@@ -1661,9 +1661,7 @@ private void getCallInfo(ref CORINFO_RESOLVED_TOKEN pResolvedToken, CORINFO_RESO
             pResult->methodFlags = FilterNamedIntrinsicMethodAttribs(pResult->methodFlags, methodToCall);
 
             var targetDetails = _compilation.TypeSystemContext.Target;
-            if (targetDetails.Architecture == TargetArchitecture.X86
-                && targetDetails.OperatingSystem == TargetOS.Windows
-                && targetMethod.IsNativeCallable)
+            if (targetDetails.Architecture == TargetArchitecture.X86 && targetMethod.IsNativeCallable)
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs]: >if [](start = 12, length = 2)

Should also update [https://github.com/dotnet/runtime/blob/8772dd31552d110aa217463c064f06b81c2a0e7d/src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs#L2921-L2932](https://github.com/dotnet/runtime/blob/8772dd31552d110aa217463c064f06b81c2a0e7d/src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs#L2921-L2932)

File: src/coreclr/src/vm/dllimportcallback.cpp (hunk 3 above)
Comment (jkotas) [reply src/coreclr/src/vm/dllimportcallback.cpp]: We always need to generate the IL stub with the GC transition now, even where there is not marshaling required. I could have kept this for non-Windows x86 but it did not seem worth the complexity. The non-Windows x86 path is pretty inefficient in general.

The comment about the CreateDelegate case is from Silverlight days. It is not that important now. That makes me think: We should allow CreateDelegate to return direct pointer to NativeCallable methods: https://github.com/dotnet/runtime/issues/34266

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs (hunk 4 above)
Comment (jkotas) [reply src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs]: Fixed

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**